Sources/Plasma/NucleusLib/pnAsyncCoreExe/pnAceSocket.cpp

```diff
@@ -0,0 +1,575 @@
+        if (err)
+            LogMsg(kLogError, "Failed to cancel socket connect operation: {}", err.message());
+    }
+}
+
+void AsyncSocketDelete(AsyncSocket conn)
+{
+    delete conn;
+}
+
+static void HardCloseSocket(AsyncSocket conn)
+{
+    asio::error_code err;
+    conn->fSock.set_option(asio::socket_base::linger(true, 0), err);
+    if (err)
+        LogMsg(kLogError, "Failed to set socket linger option: {}", err.message());
+
+    conn->fSock.close(err);
+    if (err)
+        LogMsg(kLogError, "Failed to close socket: {}", err.message());
+}
+
+void AsyncSocketDisconnect(AsyncSocket conn, bool hardClose)
+{
+    if (!conn->fSock.is_open())
+        return;
+
+    asio::error_code err;
+    conn->fSock.cancel(err);
+    if (err)
+        LogMsg(kLogError, "Failed to cancel pending operations: {}", err.message());
+
+    hsAssert(hardClose, "Only hardClose supported on sockets");
+    HardCloseSocket(conn);
+    conn->closeTimeMs |= 1;
+}
+
+static bool SocketQueueAsyncWrite(AsyncSocket conn, const void* data, unsigned bytes)
+{
+    hsLockGuard(s_connectCrit);
+    
+    // check for data backlog
+    WriteOperation * firstQueuedWrite = conn->fWriteOps.front();
+    if  (firstQueuedWrite) {
+        unsigned currTimeMs = TimeGetMs();
+        if (((long) (currTimeMs - firstQueuedWrite->queueTimeMs) >= (long) kBacklogFailMs)
+        &&  ((long) (currTimeMs - conn->initTimeMs) >= (long) kBacklogInitMs)
+        ) {
+            PerfAddCounter(kAsyncPerfSocketDisconnectBacklog, 1);
+
+            if (conn->fConnectionType) {
+                LogMsg(
+                    kLogPerf,
+                    "Backlog, c:{} q:{}, i:{}",
+                    conn->fConnectionType,
+                    currTimeMs - firstQueuedWrite->queueTimeMs,
+                    currTimeMs - conn->initTimeMs
+                );
+            }
+            AsyncSocketDisconnect((AsyncSocket) conn, true);
+            return false;
+        }
+    }
+
+    // If the last buffer still has space available then add data to it
+    if (!conn->fWriteOps.empty()) {
+        WriteOperation* op = conn->fWriteOps.back();
+        unsigned bytesLeft = std::min(op->fAllocSize - op->fNotify.bytes, bytes);
+        if (bytesLeft) {
+            PerfAddCounter(kAsyncPerfSocketBytesWaitQueued, bytesLeft);
+            memcpy(op->fNotify.buffer + op->fNotify.bytes, data, bytesLeft);
+            op->fNotify.bytes += bytesLeft;
+            data = (const uint8_t *)data + bytesLeft;
+            bytes -= bytesLeft;
+        }
+    }
+
+    if (bytes) {
+        // Allocate storage alongside the operation structure itself
+        unsigned bytesAlloc = std::max(bytes, kMinBacklogBytes);
+        auto membuf = new uint8_t[sizeof(WriteOperation) + bytesAlloc];
+        WriteOperation* op = new (membuf) WriteOperation;
+        conn->fWriteOps.emplace_back(op);
+
+        // TODO: asyncId
+        op->fAllocSize = bytesAlloc;
+        op->fNotify.asyncId = /* TODO */nullptr;

```

Comment:
Looking through the NT source trying to figure out the role of asyncId.

My understanding is that the NT sockets kept a master operation list for all read/write operations within a socket. So each op needs an id by which it can be tracked in since read/writes can happen in parallel.

The ASIO version has a dedicated array for write operations - so I'm not sure how the NT sockets work exactly maps. Buffers should always be written in order, so I should always be able to refer to the head of the write ops and assume it is the operation that just called back.

I'm still trying to better understand the NT code though. `INtOpDispatch` seems like the function that deals the most with the asyncId and the operation list.